Source/missiles.cpp

```diff
@@ -590,6 +590,90 @@ void CheckMissileCol(Missile &missile, int mindam, int maxdam, bool shift, Point
 		PlaySfxLoc(MissilesData[missile._mitype].miSFX, missile.position.tile);
 }
 
+/**
+ * @brief Could the missile collide with solid objects? (like walls or closed doors)
+ */
+bool CouldMissileCollideWithSolidObject(Point tile)
+{
+	int oid = dObject[tile.x][tile.y];
+	if (oid != 0) {
+		oid = abs(oid) - 1;
+		if (!Objects[oid]._oMissFlag)
+			return true;
+	}
+	return nMissileTable[dPiece[tile.x][tile.y]];
+}
+
+void MoveMissileAndCheckMissileCol(Missile &missile, int mindam, int maxdam, bool ignoreStart, bool ifCollidesDontMoveToHitTile)
+{
+	Point prevTile = missile.position.tile;
+	missile.position.traveled += missile.position.velocity;
+	UpdateMissilePos(missile);
+	int possibleVisitTiles = (missile.position.velocity.deltaX == 0 || missile.position.velocity.deltaY == 0) ? prevTile.WalkingDistance(missile.position.tile) : abs(prevTile.x - missile.position.tile.x) + abs(prevTile.y - missile.position.tile.y);

```

Comment:
Good thinking on checking if the missile is flying straight. I thought I'd mention that I saw the following function when I was reviewing this PR yesterday. It looks like you're using effectively the same logic in the ternary else clause.

https://github.com/diasurgical/devilutionX/blob/ec1850e2950f015d7814a8d9472bde471f4cb0d6/Source/engine/point.hpp#L142-L147

Reply:
I also think it's the same, switched to `ManhattanDistance`.
This way it looks cleaner and less code. Thanks 🙂